Integrate mesh deduplication and GPU instancing handling

alexykot opened the pull request:
_API draft proposal, not for merging_

Hi @EliCDavis 

I noticed you've added a way to enable `EXT_mesh_gpu_instancing` short after my work on mesh deduplication was merged. I'm now looking to use instancing properly on the client side, not just for file size reduction, but for faster rendering as well, and that means I need to make use of this extension.

However, this API with `PolyformModel.GpuInstances` is not compatible with the earlier `PolyformModel.T/R/S` implementation I've added, so I'm not sure what is the preferred way now. 

In this PR I'm suggesting a different API that will create a single way to control these two related features.

**Proposal**
1. Add `PolyformModel.UseGpuInstancing bool` flag, replacing `PolyformModel.GpuInstances`. This will be the only public API feature directly related to this function.
2. If the flag is set on a given model - use existing mesh deduplication mechanics to find other models that reference the same mesh.
3. Treat all models of a given mesh as enabled for GPU instancing, if any model instance has the flag set.
4. Internally build the list of instances and create a single node with one mesh reference and necessary data accessors.

**Rationale**
I see mesh deduplication  and GPU instancing as separate, but related features. Mesh deduplication and creating nodes that use references to the same mesh within a single file can be done using server-side only means, with no knowledge of the client's implementation, and has value in itself. Adding GPU instancing - builds on top of this mechanics, but expects client support of a custom (albeit very popular) extension, which is strictly speaking not part of the core format.

In the current state - these two features are disconnected and not compatible with each other. Mesh dedup implementation expects each instance to be declared as a separate model, while GPU instancing implementation requires a single model with a list of TRS values for each instance. So switching between them requires changing implementation for the library clients.

Proposed approach allows flexibility to enable the feature where necessary, but not having to materially change the logic if the feature has to be disabled if it is not supported by a given client. And it still allows using the core format feature of mesh deduplication even if the extension is not supported.

---
I assume the current state of this API is temporary, and this fork in the road will lead down one or other way. 
I'm happy with either way, but I would like to resolve this so I don't have to change my implementation again later on. 
If you're happy with my proposal - I'll create an implementation for this API change.

EliCDavis commented:
Hey! Thanks for this. Man you're really getting into the fun stuff now huh!

I think your right that the API design around this extension should change some. I think it's a great idea for the writer to take advantage of this for the user when possible,

Just for sake of keeping track of things in github's issues/PRs, here is a link to our earlier discussion on the matter.

https://github.com/EliCDavis/polyform/pull/43

# State of Affairs

## EXT_mesh_gpu_instancing Summary

https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Vendor/EXT_mesh_gpu_instancing/README.md

A few points that I will reference throughout this repsonse:

1. "designed to enable GPU instancing, rendering many copies of a single mesh at once using a small number of draw calls"
2. "As with vertex attributes, custom instance attributes may be prefixed with an underscore (e.g. _ID, _COLOR, etc.) and used for application-specific effects"
3. "Instancing only applies to mesh nodes"
4. "The attributes specify an object space transform that should be multipled by the node's world transform in the shader to produce the final world transform for the instance."
5. "Because the extension does not provide a way to specify a non-instanced fallback, files that use the extension should specify it in extensionsRequired array."
6. "The reader may notice that several desirable features are missing from the extension, such as specifying different materials for different instances within the same batch, or animation of individual instances. These are omitted, at present, for lack of a clear mechanism to support them in a sufficient number of realtime engines"

Given some of the requirements of GPU Instancing and GLTF's extension spec, we're limited in how much we can optimize. 

To satisfy item **#4**, the only time we should combine nodes into a single node for GPU instancing is when those nodes share the same parent. This isn't an issue for us today, given that a `gltf.PolyformModel` has no ability to have children. I'm not sure if it might be worth us adding children in the same pass as GPU instancing such that we can go ahead and ensure it behaves properly with the extension.

To satisfy **#6**, the only time we should combine nodes into a single node for GPU instancing is when those nodes share the same material.

## What We Already Do

* We perform a type of instancing that we call "data instancing" in which we find and keep track of duplicate references to the same `modeling.Mesh`.  
* We perform a shallow equality check on materials to de-duplicate definitions and update node references
* We perform a shallow equality check on textures to de-duplicate definitions and update meterial references.

I bring this all this up to say that in order to get the most out of GPU instancing, we need to perform the GPU instancing pass *after* we perform the 3 optimizations above. What used to be a scene with 100 meshes, materials, textures, and models might all collapse down into a scene with 1 mesh, material, and texture with 100 models. Then this GPU instancing operates on those 100 models.

## How TRS Is Currently Used

I created the TRS package to assist in transformation type operations, and simplify some of the API design. The package that probably changed the most was the `modeling/repeat` package.

One simple example of a function that changed during the API redesign was the `Line` struct. It originally would take a mesh and repeat it multiple times along a line and return a new Mesh. Now it returns a array of TRSs, which make it easier to write unit tests around, and plugs in easily to the GPU Instances API.

![image](https://github.com/user-attachments/assets/77f02eec-8aff-4fb1-a3a2-fac8297568bb)

I also use it for functionality pertaining to curves/sampling.

I bring all this to your attention to say that I at the moment the TRS package is probably the route I'm going to take for most things going forward, atleast for where it makes sense.

# Proposal

For the `gltf.Writer` definition:

* Add a new field called something like `UseGpuInstancingExt` that's a boolean.
  * If the value is true, we detect if there's any canidate models that can be instanced, then instance them and add the extension as required to the GLTF file
  * If it's false, we don't perform any optimizations

We make the extension optional to help address point **#5** in [EXT_mesh_gpu_instancing Summary](#ext_mesh_gpu_instancing-summary). If someone wanted to generate two versions of a GLTF quickly, they could just toggle this flag in the writer and write twice.

For the `gltf.PolyformModel` definition

* Rename `GPUInstances` to `Instances` and keep it type `[]trs.TRS`.
  * If `gltf.Writer.UseGpuInstancingExt` is false, then we duplicate the model for each instance provided
* Add a `InstanceCandidate` of type boolean or struct
  * If we choose boolean, then if it's true, we combine `gltf.PolyformModel` models into a single `gltf.PolyformModel` if they share the same model, material, and parent.
  * If we choose the struct route, we can define a struct that has flags that define how and when a model is valid to be combined into an instance.
    * For example, we could add further restrictions such that the "name" of the polyform model matters. So we only combine models that match in terms of name, material, mesh, and parent.
    * Another option we might consider adding to the struct is
  * I think the struct helps future proof the API, but def makes it more involved.

# Unsolved Problems

Right now, developers can only define instance data for position, rotation, and scale. Ideally, they define any arbitrary data on a per instance data to help address point **#2** in [EXT_mesh_gpu_instancing Summary](#ext_mesh_gpu_instancing-summary).

One way to address this would be to change `gltf.PolyformModel.Instances` type from `[]trs.TRS` to `modeling.Mesh` whose topology is a pointcloud. But that feels pretty ugly. I bring this up incase you had any thoughts in the manner.

---

Thanks for your patience in my response time. Work has been slamming me

Also thank you for taking the time to read this wall of text.

alexykot commented:
Ok, a few thoughts. 


1. You're suggesting to go with `gltf.PolyformModel.Instances []trs.TRS`. The `PolyformModel` currently also has:

```
Translation *vector3.Float64
Scale       *vector3.Float64
Rotation    *quaternion.Quaternion
``` 

as separate attributes, I've added those earlier and they are used currently for the data instancing implementation. I assume that indirectly you're suggesting to remove this and rely on `Instances` slice for all instancing features. 

2. I'm not sure why `PolyformModel.InstanceCandidate` needs to be exposed. This seems to be an internal implementation concern from my perspective. Actually, if we go with `Instances []trs.TRS` _only_, and remove the other `T,R,S` attributes - the data instancing will have to depend on the same `Instances` values and there will be no internal flattening happening. Effectively this will push the model deduplication concern to the client, and library will expect to receive a _single model with a list of instances_, instead of a _list of models referencing a single mesh&material_. If we go down this path - there will be no need to identify the candidates, this will be now client's job to do that.

3. I see what you mean about custom attributes, but I suggest to not confuse these two tasks. I think we need to agree the shape of instancing itself first, and with that one clear - this will be an easy question to answer.

4. `gltf.Writer` currently has no public attributes or constructor options. I'm not sure if `UseGpuInstancingExt` flag is most appropriate on the `Writer`, or maybe on the `PolyformScene`? Practically the API I interact with normally is `gltf.WriteText/WriteBinary` functions, and those accept `PolyformScene` as the main input, while writer is strictly speaking an internal concern of the implementation of those functions. I suggest adding the flag `PolyformScene.Options.UseGpuInstancingExt` instead.

5. Implementing `PolyformModel.Children` support would be a much bigger undertaking, I don't think I want to do that now. My current task is to implement GPU instancing and I want to stabilise the API and implementation for that, but not creep the scope beyond that.

---
I think the key question here is whether to use:
- one mesh, one material, many models each with unique TRS, internal model deduplication;
_OR_
- one mesh, one material, one model, many instances each with unique TRS, no internal deduplication;

The rest of the implementation will be obviously derived from the answer to that. 

I personally would prefer the former, i.e. drop the `Instances []trs.TRS` and keep `PolyformModel.T,R,S` attributes (and convert it to single `PolyformModel.Trs` attribute). This will simplify my work as my system is already set up to use "one mesh + list of models" approach, rather than "one model + list of instances", and it has no explicit concept of "instances". And correct conversion of models to node(s) is currently a library concern. 

However, I'm just a single user of the library, and I'm not sure this is a good reason to chose this API shape over the other option. So I'm happy to follow your lead on this. Which way you think we should go here?

EliCDavis commented:
I think to help discussion, it might help to bring up the fact that at the moment, the closest analog we have to a GLTF node is `PolyformModel`.

A link to the GLTF Node schema
https://github.com/KhronosGroup/glTF/blob/main/specification/2.0/schema/node.schema.json

1. Sorry, to be clear, I'm not suggesting the removal of Translation/Rotation/Scale fields on `PolyformModel`. I think those fields are important to have outside of a data instancing perspective. They help us maintain parity with GLTF's node spec, and are just generally helpful outside of any instancing context (like animations). Speaking of animation, having both T/R/S and the instances fields allows us to animate a collection of instances easily.
2. I think you make a fair point on not adding the `PolyformModel.InstanceCandidate`. I'm making an error trying to design for the future where people might want to configure instancing on a per model basis, but I really should be avoiding speculative design and just wait till someone actually opens up an issue with their specific needs. 
3. Fair enough! 
4. Not a bad idea. Also allows us to easily communicate to the developer that a scene used GPUExtension whenever I get around to creating a GLTF Reader (Speaking of, would that be useful to you? I could start on that sooner than later)
5. Fair enough!
---

The `EXT_mesh_gpu_instancing` extension is meant to have it's data specified at the node level. So to maintain parity as close as possible to the spec, as well as providing the most functionality, I think both the T/R/S fields and the Instances field should be kept.

I __also__ think we should take your suggestion, when `PolyformScene.Options.UseGpuInstancingExt` is true, automatically build a model with instances where possible by utilizing the existing T/R/S fields. I don't see any reason a user wouldn't want "free" optimizations if they've explicitly set `PolyformScene.Options.UseGpuInstancingExt` to true.

If you want to replace the T/R/S fields with a pointer to a TRS struct, I think that's valid, but I wouldn't remove the concept all together.

With this, instancing can happen two different ways:

 * implicitly as an optional writer optimization
 * explicitly called out by the user

---

So to restate explicit desired API changes:

1. `PolyformModel.GPUInstances` gets renamed to `PolyformModel.Instances`
2. `PolyformScene` Get's a new `Options` field and struct containing a `UseGpuInstancingExt` boolean field.
3. **(Optional)** `PolyformModel`'s `Transform`, `Rotation`, and `Scale` Fields are replaced with a field called `Transform` which is a pointer to a `TRS`.

alexykot commented:
Sorry, I was quite busy with other things recently. 

Ok, so I'm confused - if we have `PolyformModel.T/R/S` _and_ we have `PolyformModel.Instances` - how you propose these two should interact? I.e. if there are instances defined and there are multiple `PolyformModel`s that also can be deduplicated and collapsed into instances as well - how these two should interact in your view?

Btw, I will certainly replace T/R/S fields with a struct. I'm not sure why I haven't done this originally, but now that's pretty obvious thing to do imo.

EliCDavis commented:
You can see how T/R/S interacts with Instances in the example below. The same Instances/meshes are used, but one node has an offset.

![image](https://github.com/user-attachments/assets/7885f751-7974-48f7-8bb1-6c81b85d2a0e)

---

Going back to the extension's spec:

https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Vendor/EXT_mesh_gpu_instancing/README.md

1. "Instancing only applies to mesh nodes"
2. "The attributes specify an object space transform that should be multipled by the node's world transform in the shader to produce the final world transform for the instance."
3. "The reader may notice that several desirable features are missing from the extension, such as specifying different materials for different instances within the same batch, or animation of individual instances. These are omitted, at present, for lack of a clear mechanism to support them in a sufficient number of realtime engines"

Given some of the requirements of GPU Instancing and GLTF's extension spec, we're limited in how much we can optimize.

To satisfy item # 2, the only time we should combine nodes into a single node for GPU instancing is when those nodes share the same parent. This isn't an issue for us today, given that a gltf.PolyformModel has no ability to have children.

To satisfy # 3, the only time we should combine nodes into a single node for GPU instancing is when those nodes share the same material.

---

To combine two sets of instances together, we get to do some matrix math:

Let's say we have a model `A` with instance `X`, and model `B` with instance `Y`. We want to create a new model `C` that has 2 instances `X'` and `Y'`

We create a model C with an identity matrix for it's TRS. This corresponds to no rotation, a scale of 1, and no translation.

Then it's two instances are:
`A * X = X'`
`B * Y = Y'`

alexykot commented:
Yes, deduplication only happens if all properties apart from the TRS are the same, i.e. same mesh, same material, same samplers (`name` is ignored tho). That's a prerequisite for to this to be even considered.

Ok, cutting to the chase - you're suggesting to have both mechanics' in place, so the explicitly defined Instances and implicitly identified model copies will be both flattened into a list of `GPU_instances` on serialization.

Hmm, why? 
From the perspective of API design - why would we want to have two unrelated separate ways of achieving exactly the same outcome? This does complicate the API and the internal implementation, but what is the added value here?

Btw, my perspective as a library user - I'm now aware of the node mechanics of the GLTF spec, but only because I had to dabble in `polyform` itself to implement optimizations I needed. I imagine other users would not have to do that, so they may not be aware of this mechanics at all, and I think this is fine. I assume the purpose of an API is to simplify and abstract internal details, and reduce exposed complexity. Keeping two separate ways of achieving the same goal goes against that purpose, so I'm not sure what other purpose this would be serving. So what is that purpose?

EliCDavis commented:
To me, there are pros to having both ways:

### Pros For Explicit Declaration of Instance

1. If I already have all the TRSs that I want instanced, it's more work for me to duplicate the models, then it's also more work for the writer to de-duplicate the models I just made
2. It's an explicit opt-in to using the extension. 
3. It's the only way to define animation data on a group of instances. Conceptually, it's impossible implicitly (until the concept of children is added)
4. Faster writing speeds

### Pros For Implicit Detection of Instances

A user doesn't have to think about it.

### Issues With Implicit Detection

The author of the model still needs to be able to opt into the extension. The instancing extension isn't implemented everywhere. For example, Window's Model Viewer doesn't have it implemented. If the writer just went ahead and auto-turned on the extension's usage, people might wonder why their GLB looks broken one way, but if they change a material on one of the models, the GLB all of a sudden looks correct (because we're now unable to instance). 

So even if implicit detection is implemented, there still needs to be some "opt-in" to the situation.

--- 

I am not convinced polyform should part with an explicit declaration of instances. 

I like the idea of the writer _also_ helping the user out by detecting instances "for free," but that's def a nice to have. If you don't think it makes sense, I'm okay with not having it at the moment.

alexykot commented:
Okay, so now I had a reason to come back to this and actually get this done and so I did. 

Here's what's done.
- added flag `PolyformScene.UseGpuInstancing` - if set, this will enable instancing for both explicitly defined GPU instances of the models and models deduplicated implicitly. This flag is required if `GpuInstances` are defined for any models.
- Internally, the models are now deduplicated with consideration of instances and a single list of instantiable models is created. Effectively, this means that multiple models with individual instances may be collapsed into a single list if all the details match.
- If a model has any kind of animations attached, it will not be collapsed and will always be a separate node. It may still use the `GpuInstancing` if all other requirements are met, and then all instances will share the same animation.

This integrates the support for implicit model deduplication and explicit setting of GPU instances into an orchestrated system. If `UseGpuInstancing` is enabled, it will be used to the max. If it is not enabled, each model will have its own node, but it will still reference the same mesh if possible.

@EliCDavis Let me know what you think about this API and this approach to solution.

---
Also, while testing this I discovered some issues with offsets for vector data and I had to update the alignment for binary offsets. They seem to have updated a few tests, but haven't broken anything I think, and it fixed the test files with `GpuInstancing` data that otherwise could not be rendered.

---
Ironically, after I've implemented all this and tried to test it on my system, I found out that Draco compressor that I'm using is incompatible with `EXT_mesh_gpu_instancing` extension. Between the two I choose Draco, so I will not be able to use this solution right now.

EliCDavis commented on formats/gltf/writer.go:
```diff
@@ -339,88 +357,175 @@ func rgbToFloatArr(c color.Color) [3]float64 {
+			}
 
-			newNode.Extensions[extGpuInstancingID] = instances
+			newNode.Translation = &posArr
```
could we do same cheks here like we do the rotation? only set position if it's not (0,0,0), and only set scale if it's not (1,1,1)

alexykot commented:
Oh, you merged this already. I was going to get back to this. It was sitting in my inbox for a few days. 
Then I guess it's ok :)
